Package ESMValTool

zklaus opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`
- ruby `@conda-forge/help-ruby`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams. You can [ping the team][1] using a special command in
a comment on the PR to get the attention of the `staged-recipes` team. You can
also consider asking on our [Gitter channel][2] if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 1 to supersede the earlier build of https://github.com/conda-forge/esmvaltool-python-feedstock.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/esmvaltool```) and found some lint.

Here's what I've got...


For **recipes/esmvaltool**:

 * The recipe could do with some maintainers listed in the `extra/recipe-maintainers` section.
 * The recipe `license` should not include the word "License".
 * ``requirements: run: python>=3.7`` must contain a space between the name and the pin, i.e. ``python >=3.7``
 * ``requirements: run: iris>=3.0.1`` must contain a space between the name and the pin, i.e. ``iris >=3.0.1``

For **recipes/esmvaltool**:

 * License is not an SPDX identifier (or a custom LicenseRef) nor an SPDX license expression.

Documentation on acceptable licenses can be found [here]( https://conda-forge.org/docs/maintainer/adding_pkgs.html#spdx-identifiers-and-expressions ).

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/esmvaltool```) and found it was in an excellent condition.

zklaus commented:
tl;dr: Can rename conda-forge/esmvaltool-python-feedstock to esmvaltool-feedstock?


@conda-forge/staged-recipes, I would like to request some guidance on how to move on. The situation is as follows:

The project is principally a Python package for the analysis of climate models. It supports user written diagnostics in several languages, at the moment Python, R, NCL, and Julia. The packaging is now realized as a multi-output where `esmvaltool-python` is the main package containing both the tool itself and the Python diagnostics, separate packages for the R and NCL diagnostics, called `esmvaltool-r` and `esmvaltool-ncl`, respectively, and an overall meta package, `esmvaltool`, that depends on all of these.
In other words, `esmvaltool` depends on `esmvaltool-python`, `esmvaltool-r`, and `esmvaltool-ncl`; `esmvaltool-r` depends on `esmvaltool-python`; `esmvaltool-ncl` depends on `esmvaltool-python`.

Because that is the principal package, we started with the esmvaltool-python-feedstock, with the intention of adding esmvaltool-r-feedstock, esmvaltool-ncl-feedstock, and esmvaltool-feedstock.
However, with the successful integration of the multi-output recipe in esmvaltool-python-feedstock (onda-forge/esmvaltool-python-feedstock#6), the other feedstocks are no longer necessary, so ideally we would rename esmvaltool-python-feedstock to esmvaltool-feedstock. Is this possible?

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/esmvaltool```) and found it was in an excellent condition.


I do have some suggestions for making it better though...


For **recipes/esmvaltool**:

 * It looks like the 'esmvaltool' output doesn't have any tests.
 * A conda package with same name (esmvaltool-python) already exists.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/esmvaltool```) and found it was in an excellent condition.


I do have some suggestions for making it better though...


For **recipes/esmvaltool**:

 * A conda package with same name (esmvaltool-python) already exists.

zklaus commented:
After a discussion on gitter with @tdegeus and @beckermr, we decided to continue with this PR for a single esmvaltool-feedstock with a multi-output build as developed in conda-forge/esmvaltool-python-feedstock#6 and subsequent retirement of esmvaltool-python-feedstock.

That is why the build-number is 1, not 0, because it replaces the earlier esmvaltool-python build. This is also the reason why the linter complains about an existing package with the same name.

zklaus commented:
This is a noarch package that is failing on windows due to a missing dependency. This should not stop us from merging, c.f. #14700.

zklaus commented:
@conda-forge/esmvaltool-python, could you guys please have a look? This is the same multi-output recipe that we have been developing in conda-forge/esmvaltool-python-feedstock#6, only in a new feedstock (esmvaltool-feedstock), plus the following changes:

- I removed most of the pins; now the pinning is as in upstream
- I added basic testing for the top-level (`esmvaltool`) package
- I added the `console_scripts` entry points to comply with [the noarch requirements](https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-python)
- Other minor changes (formatting etc)

If you agree, I think this can go in. The Linux build works, though both osx and windows build here fail due to missing or unresolvable dependencies. Note that this does not prevent one form trying to install it on osx or windows since it will be built as noarch. Then we can try to improve the situation by working on the missing dependencies on those platforms.

valeriupredoi commented:
@zklaus very many thanks for battling and winning the fight with our dependencies list and making sure this works! I am very happy to see this through (gonna approve after I post this message), since the linux-64 distribution works fine, all's good by me. Note, however, that we may get a kick in build times and (probably) better deps solving by retiring Python 3.7, our tests have already started failing badly at figuring an env with it on [GA](https://github.com/ESMValGroup/ESMValTool/runs/3267056098?check_suite_focus=true). Cheers, mate :beer:

bouweandela commented on recipes/esmvaltool/yum_requirements.txt:
```diff
@@ -0,0 +1 @@
+xorg-x11-server-Xorg
```
Why do we need x11? I'm pretty sure our users are running the tool on systems without a graphical user interface.

zklaus replied:
Removing x11 dependencies is exactly why we need this. It came in as part of conda-forge/esmvaltool-python-feedstock#1 when we moved from depending on `matplotlib` to `matplotlib-base` which removes the `pyqt` dependency and a whole bunch of x11 dependencies. To still be able to build, we need this yum requirement on the build machine.

bouweandela commented:
It looks like the `esmvaltool-python` package is building fine for OSX, but failing to solve the dependencies for `esmvaltool-ncl`. In our documentation we say that `esmvaltool-python` can be installed on OSX, see [here](https://docs.esmvaltool.org/en/latest/quickstart/installation.html#installation-on-macosx). Would it be possible to build just the `esmvaltool-python` package for OSX now, so that remains available for our OSX users?

zklaus commented:
> It looks like the `esmvaltool-python` package is building fine for OSX, but failing to solve the dependencies for `esmvaltool-ncl`. In our documentation we say that `esmvaltool-python` can be installed on OSX, see [here](https://docs.esmvaltool.org/en/latest/quickstart/installation.html#installation-on-macosx). Would it be possible to build just the `esmvaltool-python` package for OSX now, so that remains available for our OSX users?

Everything is marked as `noarch` which means that once this PR is merged and esmvaltool has its own feedstock, things will be build only once (on a Linux machine), but will be installable on any of the platforms. Of course, where missing dependencies prevent a successful installation, the installation will then fail, but installing `esmvaltool-python` on osx should not be a problem.

bouweandela commented:
Thanks @zklaus, let's get this merged!

zklaus commented:
@dopplershift, since you already helped with #14700 and #14400, would you mind having a look? This PR is in part designed to replace #14700. Sorry for the confusion.

beckermr commented:
All of the maintainers in the recipe need to comment that they are ok being on it or they need to be removed.

zklaus commented:
@valeriupredoi, @nielsdrost, @jvegasbsc, @bouweandela could you guys please comment here, confirming that you are willing to co-maintain this package?

bouweandela commented:
I'm OK with being on it, thanks for adding me.

nielsdrost commented:
Yes, no problem for being on the list of co-maintainers, thanks for the effort!

valeriupredoi commented:
cheers for adding me @zklaus and of course (my horse) I am OK with being a maintainer. Sorry for the delayed reply, I was away :+1: :beer:

zklaus commented:
@beckermr, thanks for your help. I have removed one maintainer and gotten the comments from all the remaining ones (see above).
Do you have any other comments?

zklaus commented:
Thanks, @beckermr :tada: !

zklaus commented:
> I'm OK with being on it, thanks for adding me. And sorry for the late reply

Thanks Javi! Unfortunately, it's too late for this PR, but we will add you in an update on the new feedstock.
